mgt-file-list (depends on mgt-file)

beth-panx opened the pull request:
### PR Type
 - Feature #927
 
### Description of the changes
Add mgt-file-list component.

Originally branched off from main. So there are some commits that are already merged into next/file from bethpan/file.

**Progress checklist**
- [x] Render file by file type p0
- [x] Dark theme p0
- [x] Paging to show more based on graph result pages
- [ ] Cache p2
- [ ] Search query p1

| Query | Use if | Permission Scopes |
| ----- | ------ | ----------------- |
|`POST /beta/search/query` Request body: `{"requests": [{"entityTypes": ["driveItem", "listItem"], "query":{"queryString":"foo"}}]}` | `search-query` | Files.Read, Files.Read.All, Files.ReadWrite.All, Sites.Read.All, Sites.ReadWrite.All |


### PR checklist
- [x] Project builds (`yarn build`) and changes have been tested in supported browsers
- [x] All public classes and methods have been documented
- [ ] Added appropriate [documentation](https://github.com/microsoftgraph/microsoft-graph-docs/tree/master/concepts/toolkit) [target branch `mgt/next` for new features]. Docs PR: <!-- Link to docs PR here -->
- [x] License header has been added to all new source files (`yarn setLicense`)
- [x] Contains **NO** breaking changes

nmetulev commented on packages/mgt-components/src/components/mgt-file/mgt-file.ts:
```diff
@@ -555,6 +562,7 @@ export class MgtFile extends MgtTemplatedComponent {
     // evaluate to true when only item-id or item-path is provided
     const getFromMyDrive = !this.driveId && !this.siteId && !this.groupId && !this.listId && !this.userId;
 
+
```
```suggestion
```

beth-panx commented:
@nmetulev @amrutha95 @vogtn feel free to review now. I did a design pass with Ying and the only thing that needs more work is the 'show-more' button. I will update that later. All the requested changes from initial review should be included by now.

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+  getUserInsightsFiles
+} from '../../graph/graph.files';
+
+import { OfficeGraphInsightString, ViewType } from '../../graph/types';
```
I should have asked this in the file PR, what do you think about renaming this to `InsightType` instead of `OfficeGraphInsightString`?

beth-panx replied:
The interface for insights are as this. I wanted to follow the pattern from graph entity. Let me know what you think
```
export interface OfficeGraphInsights extends Entity {
    // Access this property from the derived type itemInsights.
    shared?: NullableOption<SharedInsight[]>;
    // Access this property from the derived type itemInsights.
    trending?: NullableOption<Trending[]>;
    // Access this property from the derived type itemInsights.
    used?: NullableOption<UsedInsight[]>;
}
```

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+    attribute: 'load-more-file-count',
+    type: Number
+  })
+  public loadMoreFileCount: number;
```
Consider renaming this to `pageSize` (`page-size`)

beth-panx replied:
I love it. I couldn't think of a good name haha.

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+    super();
+
+    this.showMax = 5;
+    this.loadMoreFileCount = 5;
```
10 seems like a more standard page size for a list of items

```suggestion
    this.loadMoreFileCount = 10;
```

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+  constructor() {
+    super();
+
+    this.showMax = 5;
```
I'm wondering if we need `showMax`? Would we be able to accomplish the same by using `pageSize` and not calling for more items?

beth-panx replied:
Depends if the initial load amount is always the same as page size. I like the idea of having both options. This also follows what we already have with other toolkit components.

beth-panx replied:
Removed show-max

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+      }
+    }
+  })
+  public view: ViewType;
```
What do you think about `itemView` instead of `view` since it's setting the view of each item?

beth-panx replied:
Sure thing!

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+  protected renderMoreFileButton(): TemplateResult {
+    const extra = this.files.length - this._renderedFileCount;
+    return html`
+        <li class="show-more" @click=${() =>
```
Could we use a button here instead of the `li` directly?

beth-panx replied:
I am updating this to <a>

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+        }
+      }
+
+      this.files = files;
+
+      // filter files when extensions are provided
+      let filteredByFileExtension: MicrosoftGraph.DriveItem[];
+      if (this.fileExtensions && this.fileExtensions !== null) {
+        filteredByFileExtension = files.filter(file => {
+          for (const e of this.fileExtensions) {
+            if (e == this.getFileExtension(file.name)) {
+              return file;
+            }
+          }
+        });
+      }
+
+      if (filteredByFileExtension && filteredByFileExtension.length >= 0) {
+        this.files = filteredByFileExtension;
+      }
```
small optimization - moving setting the `this.files` only if there are no filtered files. This avoids setting the `this.files` property twice and causing a re-render between filtering.

```suggestion
      // filter files when extensions are provided
      let filteredByFileExtension: MicrosoftGraph.DriveItem[];
      if (this.fileExtensions && this.fileExtensions !== null) {
        filteredByFileExtension = files.filter(file => {
          for (const e of this.fileExtensions) {
            if (e == this.getFileExtension(file.name)) {
              return file;
            }
          }
        });
      }

      if (filteredByFileExtension && filteredByFileExtension.length >= 0) {
        this.files = filteredByFileExtension;
      } else {
        this.files = files;
      }
```

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+  private _fileExtensions: string[];
+  private _userId: string;
+  private _renderedFileCount: number;
+  private _filesToRender: MicrosoftGraph.DriveItem[];
```
You can make this an @internalProperty and avoid having to call requestUpdate - when the property changes, it will rerender

```suggestion
  @internalProperty() private _filesToRender: MicrosoftGraph.DriveItem[];
```

nmetulev commented on packages/mgt-components/src/graph/graph.files.ts:
```diff
@@ -201,3 +201,299 @@ export async function getUserInsightsDriveItemById(
+  }
+}
+
+async function getFilesPageIterator(graph: IGraph, request) {
```
This function should return the page iterator instead of fetching all pages. The file list should use the next() method on the page iterator to fetch additional items as needed.

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,592 @@
+    }
+  }
+
+  private showMoreFiles(count: number) {
```
Update this function to call the graph for additional  items - currently all files are fetched at once and this can cause performance issues for a large list of files

beth-panx commented:
All feedback are addressed. 👯 I feel like I didn't add the loading spinner correctly... had a hard time testing it. Feel free to modify that piece or make suggestions!

nmetulev commented on packages/mgt-components/src/graph/graph.files.ts:
```diff
@@ -201,3 +201,427 @@ export async function getUserInsightsDriveItemById(
+  let filesPageIterator;
+
+  try {
+    if (top) {
+      request = graph
+        .api(`/drives/${driveId}/root:/${itemPath}:/children`)
+        .top(top)
+        .middlewareOptions(prepScopes(scopes));
+    } else {
+      request = graph.api(`/drives/${driveId}/root:/${itemPath}:/children`).middlewareOptions(prepScopes(scopes));
+    }
```
You can simplify this here (and everywhere else in this file where you use this pattern), so you don't repeat the same query twice

```suggestion
    request = graph.api(`/drives/${driveId}/root:/${itemPath}:/children`).middlewareOptions(prepScopes(scopes));
    if (top) {
      request.top(top);
    }
```

beth-panx replied:
I can update this!

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,606 @@
+ * @class MgtFileList
+ * @extends {MgtTemplatedComponent}
+ *
+ * @cssprop
```
missing `@cssprop`s and `@fires` for events

beth-panx replied:
Good catch

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,606 @@
+    }
+
+    this._files = value;
+    this.requestStateUpdate();
```
You don't need to call `requestStateUpdate` here (in fact, you should get rid of the getter/setter for `files`) since changing this prop doesn't require state update (calling the graph). This will cause the the `loadState` method to be called each time the files are set.

beth-panx replied:
What if a developer wants to set files from js? Does it update correctly without a getter/setter?

nmetulev replied:
It does, the `@property` decorator autogenerates the setters and getters if you don't provide them

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,606 @@
+
+    const getFromMyDrive = !this.driveId && !this.siteId && !this.groupId && !this.userId;
+
+    if (
+      (this.driveId && (!this.itemId && !this.itemPath)) ||
+      (this.groupId && (!this.itemId && !this.itemPath)) ||
+      (this.siteId && (!this.itemId && !this.itemPath)) ||
+      (this.userId && (!this.insightType && (!this.itemId && !this.itemPath)))
+    ) {
+      this.files = null;
+    }
```
This could use comments on why the state is cleared here

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,606 @@
+      let filteredByFileExtension: MicrosoftGraph.DriveItem[];
+      if (this.fileExtensions && this.fileExtensions !== null) {
+        // retrive all pages before filtering
+        if (pageIterator && pageIterator.value) {
+          files = pageIterator.value;
+
+          while (pageIterator.hasNext) {
+            await pageIterator.next();
+            files = pageIterator.value;
+          }
```
```suggestion
        if (pageIterator && pageIterator.value) {
          while (pageIterator.hasNext) {
            await pageIterator.next();
          }
          files = pageIterator.value;
```

beth-panx replied:
Will update

shweaver-MSFT commented:
When I hover over an item, I see either the traditional arrow or the text pipe. Might be nice to use `cursor: pointer` to show a hand icon instead indicating that the item is clickable.

shweaver-MSFT commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,606 @@
+    const getFromMyDrive = !this.driveId && !this.siteId && !this.groupId && !this.userId;
+
+    if (
+      (this.driveId && (!this.itemId && !this.itemPath)) ||
```
I see this a few times: `(!this.itemId && !this.itemPath)`. Might be nice to de-duplicate:
```
if (!this.itemId && !this.itemPath && (this.driveId || this.groupId || this.siteId || (userId && !this.insightType)) {
```

beth-panx replied:
I agree with you for code simplicity. I do think the current implementation if more readable in understanding why those are listed as is. These are the requirements of the Api endpoint of what combination of attributes are needed.

shweaver-MSFT replied:
Gotcha, I didn't realize there was a strong correlation. Might be worth commenting that :)

beth-panx replied:
Yes! Adding comments now.

shweaver-MSFT commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,606 @@
+      } else if (this.fileQueries && this.fileQueries !== null) {
+        files = await getFilesByQueries(graph, this.fileQueries);
+      } else if (getFromMyDrive) {
+        if (this.itemId && this.itemId !== null) {
```
This is a little redundant. If `this.itemId` evaluates as true, it is never going to be null.

beth-panx replied:
If you have item-id added as an attribute but not add any value, storybook hangs.

shweaver-MSFT replied:
I'm confused. What value does storybook set it to? My point is, if you are simply detecting truthiness of the attribute, the first check will do it. `this.whatever` will return true for any non-null value, so the `!== null` check is redundant.

beth-panx replied:
You can supply a component as this: `<mgt-file-list item-id="" />` and `itemId` exists but evaluates to `null`. `this.itemId` is true.

beth-panx replied:
Yeah javascript has a weird definition on null vs. undefined. And typeof null is an object. 🤣🙈🤷🏻‍♀️

shweaver-MSFT replied:
Fricken JS! Ok, ya there is definitely some weirdness here. It bothered me so I had to test a bit. Storybook aside, passing an empty attribute `item-id=""` causes the value to reflect as an empty string object `""`.

Because JS is weird, the evaluation of `(this.itemId && this.itemId !== null)` results as... also an empty string `""`; the value of the first evaluated object, not type-coerced to a bool as I expected.

So in the end, we are really just evaluating the empty string as truthy/falsy followed by a literal null check.

Imo, this is nonobvious behavior and I think it needs some sort of commenting or be tweaked a bit. Using double negation `!!` would make it more clear to me because I can tell that we are explicitly evaluating the truthiness of the attribute `if (!!this.itemId) {` 

Thoughts?

shweaver-MSFT commented on packages/mgt-components/src/components/mgt-file/mgt-file.ts:
```diff
@@ -620,17 +628,26 @@ export class MgtFile extends MgtTemplatedComponent {
         case 'size':
           // convert size to mb
-          const sizeInMb = (driveItem.size / (1024 * 1024)).toFixed(2);
+          let sizeInMb;
```
Thoughts on adding support for KB or GB?

beth-panx replied:
Hmm I did this to follow design. It's a good point. Maybe we should. Will look at this shortly.

beth-panx replied:
I updated this to convert to kb, mb, gb... but it will be in the mgt-file.ts

shweaver-MSFT commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,606 @@
+   * @returns
+   * @memberof MgtFileList
+   */
+  protected async loadState() {
```
Need to implement `clearState` as well to clear out any values set here.

shweaver-MSFT commented:
I'm not sure on the visual here, is that grey bar part of the design?

![image](https://user-images.githubusercontent.com/25832310/112230701-8be02080-8bf2-11eb-9794-0ad4eb943e08.png)

beth-panx commented:
> I'm not sure on the visual here, is that grey bar part of the design? It appears on hover for me
> 
> ![image](https://user-images.githubusercontent.com/25832310/112230701-8be02080-8bf2-11eb-9794-0ad4eb943e08.png)

Yeah I actually just saw this and updated css on it. Should remove those 6px at the bottom in the next commit.

shweaver-MSFT commented:
I noticed the control doesn't seem to respond if I change attribute values at runtime. Might need to listen to the filelist's own property change events to detect when certain properties are changed and clear/reload state as appropriate.

shweaver-MSFT reviewed (changes requested):
Left a few comments. The big one for me is the ability to change attributes at runtime and have the control respond gracefully. 

You could try doing this in the attribute setters like MgtPerson, but there are so many attributes in MgtFileList that I suggest overriding `attributeChangedCallback` instead. MgtPersonCard has a good example of how to do this: https://github.com/microsoftgraph/microsoft-graph-toolkit/blob/e4ae102d409df112a48878050061b10e3c81c19d/packages/mgt-components/src/components/mgt-person-card/mgt-person-card.ts#L319

beth-panx commented:
@nmetulev Addressed all your comments. Please take a look.
@shweaver-MSFT Addressed all your comments. Added `clearState()` and `attributeChangedCallback()`. Would love your thoughts on implementation.

nmetulev commented:
Rather than implementing `attributeChangedCallback` which only works for attributes and not properties (unless they are reflected), I recommend calling `clearState()` in the setters right before calling `requestStateUpdate()`

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -0,0 +1,692 @@
+    this.itemView = ViewType.twolines;
+  }
+
+  /**
+   * Synchronizes property values when attributes change.
+   *
+   * @param {*} name
+   * @param {*} oldValue
+   * @param {*} newValue
+   * @memberof MgtPersonCard
+   */
+  public attributeChangedCallback(name: string, oldValue: string, newValue: string) {
+    super.attributeChangedCallback(name, oldValue, newValue);
+
+    if (oldValue === newValue) {
+      return;
+    }
+
+    switch (name) {
+      case 'file-list-query':
+      case 'file-queries':
+      case 'site-id':
+      case 'drive-id':
+      case 'group-id':
+      case 'item-id':
+      case 'item-path':
+      case 'user-id':
+      case 'insight-type':
+        this.clearState;
+        this.requestStateUpdate();
+        break;
+    }
+  }
```
```suggestion
  protected requestStateUpdate(force?: boolean) {
    this.clearState();
    return super.requestStateUpdate(force);
  }
```

This should work to handle clearing the state when you need to request a state update. The `attributeChangedCallback` doesn't always get called when a property is changed because the underling attribute for a property is not always updated.

@shweaver-MSFT, what do you think about updating the baseComponent `requestStateUpdate` to call `clearState` by default - it looks like most components already overwrite this method to clearState anyway.

shweaver-MSFT replied:
Sounds good to me, but let's save that for a separate PR :)

nmetulev reviewed (changes requested):
* I'd expect the background color for click state to change based on the designs - currently it is the same as hover state
* I don't see a loading spinner when I click on show more - it just hangs until more items are loaded
* styling:
  * it doesn't look like I can change/remove the border, dropshadow, and margins - I expect this would be needed by most devs
  * there doesn't seem to be a way to style the show more button
  * add a story with all different styles - this can make testing the different css props very easy. 
* add a story for testing `itemClick` event

beth-panx commented:
@nmetulev @shweaver-MSFT I think I am done haha

shweaver-MSFT reviewed (approved):
You got all my comments :) Looks good

nmetulev commented on stories/components/fileList.stories.js:
```diff
@@ -0,0 +1,134 @@
+  <mgt-file-list disable-expansion></mgt-file-list>
+`;
+
+export const selectedFile = () => html`
+  <p>Selected File:</p>
```
```suggestion
export const itemClickEvent= () => html`
  <p>Clicked File:</p>
```

nmetulev reviewed (commented):
Amazing - few minor things, but overall looks good and can be merged

beth-panx commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -350,6 +350,17 @@ export class MgtFileList extends MgtTemplatedComponent {
+   * @memberof MgtFileList
+   */
+  @property({
+    attribute: 'disable-expansion',
```
Yes can do 👯

beth-panx commented:
I will write up the documentation for file and file-list today or tomorrow :)

nmetulev commented on packages/mgt-components/src/components/mgt-file-list/mgt-file-list.ts:
```diff
@@ -356,14 +347,13 @@ export class MgtFileList extends MgtTemplatedComponent {
     type: Boolean
   })
-  public disableExpansion: boolean;
+  public hideMoreFileButton: boolean;
```
a nitpick, could we make it plural? `hideMoreFilesButton`

nmetulev reviewed (approved):
🥳
